orderbook/src/orderbook.rs

```diff
@@ -0,0 +1,89 @@
+use model::{Order, UserOrder};
+use primitive_types::H160;
+use tokio::sync::RwLock;
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum AddOrderError {
+    AlreadyExists,
+    InvalidSignature,
+    #[allow(dead_code)]
+    PastNonce,
+    #[allow(dead_code)]
+    PastValidTo,
+}
+
+#[derive(Debug)]
+pub enum RemoveOrderError {
+    DoesNotExist,
+}
+
+#[derive(Debug, Default)]
+pub struct OrderBook {
+    // TODO: Store more efficiently (for example HashMap) depending on functionality we need.
+    pub orders: RwLock<Vec<Order>>,
+}
+
+impl OrderBook {
+    pub async fn add_order(&self, order: UserOrder) -> Result<(), AddOrderError> {
+        // TODO: Check order signature, nonce, valid_to.
+        let mut orders = self.orders.write().await;
+        if orders.iter().any(|x| x.user_provided == order) {
+            return Err(AddOrderError::AlreadyExists);
+        }
+        let order = user_order_to_full_order(order).ok_or(AddOrderError::InvalidSignature)?;
+        orders.push(order);
+        Ok(())
+    }
+
+    pub async fn get_orders(&self) -> Vec<Order> {
+        self.orders.read().await.clone()
+    }
+
+    #[allow(dead_code)]
+    pub async fn remove_order(&self, order: &UserOrder) -> Result<(), RemoveOrderError> {
+        let mut orders = self.orders.write().await;
+        if let Some(index) = orders.iter().position(|x| x.user_provided == *order) {
+            orders.swap_remove(index);
+            Ok(())
+        } else {
+            Err(RemoveOrderError::DoesNotExist)
+        }
+    }
+}
+
+/// None if signature doesn't match.
+fn user_order_to_full_order(user_order: UserOrder) -> Option<Order> {
```

Comment:
nit: I find having this return None when the signature doesn't match a bit implicit (it requires reading the comment). Maybe we can recover the signtature in a separate method (where it's clear that None means invalid singature) and then have this method take the owner as an argument?